src/constants/volumeRayMarchShader.ts

```diff
@@ -1,297 +1,293 @@
-  "  col = luma2Alpha(col, GAMMA_MIN, GAMMA_MAX, GAMMA_SCALE);",
-  "return col;",
-  "}",
-
-  "bool intersectBox(in vec3 r_o, in vec3 r_d, in vec3 boxMin, in vec3 boxMax,",
-  "                  out float tnear, out float tfar){",
+  retval.rgb *= maskVal;
+  return bounds*retval;
+}
+
+bool intersectBox(in vec3 r_o, in vec3 r_d, in vec3 boxMin, in vec3 boxMax,
+                  out float tnear, out float tfar) {
   // compute intersection of ray with all six bbox planes
-  "  vec3 invR = vec3(1.0,1.0,1.0) / r_d;",
-  "  vec3 tbot = invR * (boxMin - r_o);",
-  "  vec3 ttop = invR * (boxMax - r_o);",
+  vec3 invR = vec3(1.0,1.0,1.0) / r_d;
+  vec3 tbot = invR * (boxMin - r_o);
+  vec3 ttop = invR * (boxMax - r_o);
 
   // re-order intersections to find smallest and largest on each axis
-  "  vec3 tmin = min(ttop, tbot);",
-  "  vec3 tmax = max(ttop, tbot);",
+  vec3 tmin = min(ttop, tbot);
+  vec3 tmax = max(ttop, tbot);
 
   // find the largest tmin and the smallest tmax
-  "  float largest_tmin  = max(max(tmin.x, tmin.y), max(tmin.x, tmin.z));",
-  "  float smallest_tmax = min(min(tmax.x, tmax.y), min(tmax.x, tmax.z));",
+  float largest_tmin  = max(max(tmin.x, tmin.y), max(tmin.x, tmin.z));
+  float smallest_tmax = min(min(tmax.x, tmax.y), min(tmax.x, tmax.z));
 
-  "  tnear = largest_tmin;",
-  "  tfar = smallest_tmax;",
+  tnear = largest_tmin;
+  tfar = smallest_tmax;
 
   // use >= here?
-  "  return(smallest_tmax > largest_tmin);",
-  "}",
-
-  "vec4 accumulate(vec4 col, float s, vec4 C) {",
-  "    float stepScale = (1.0 - powf((1.0-col.w),s));",
-  "    col.w = stepScale;",
-  "    col.xyz *= col.w;",
-  "    col = clamp(col,0.0,1.0);",
-
-  "    C = (1.0-C.w)*col + C;",
-  "    return C;",
-  "}",
-
-  "vec4 accumulateMax(vec4 col, float s, vec4 C) {",
-  "    if (col.x*col.w > C.x) { C.x = col.x*col.w; }",
-  "    if (col.y*col.w > C.y) { C.y = col.y*col.w; }",
-  "    if (col.z*col.w > C.z) { C.z = col.z*col.w; }",
-  "    if (col.w > C.w) { C.w = col.w; }",
-  "    return C;",
-  "}",
-
-  "vec4 integrateVolume(vec4 eye_o,vec4 eye_d,",
-  "                     float tnear,   float tfar,",
-  "                     float clipNear, float clipFar,",
-  "                     sampler2D textureAtlas",
-  "                     ){",
-  " vec4 C = vec4(0.0);",
-  " float tend   = tfar;",
-  " float tbegin = tnear;",
-
-  //'  // march along ray from front to back, accumulating color',
-
-  //'  //estimate step length',
-  "  const int maxSteps = 512;",
+  return(smallest_tmax > largest_tmin);
+}
+
+vec4 accumulate(vec4 col, float s, vec4 C) {
+  float stepScale = (1.0 - powf((1.0-col.w),s));
+  col.w = stepScale;
+  col.xyz *= col.w;
+  col = clamp(col,0.0,1.0);
+
+  C = (1.0-C.w)*col + C;
+  return C;
+}
+
+vec4 accumulateMax(vec4 col, float s, vec4 C) {
+  if (col.x > C.x) { C.x = col.x; }
+  if (col.y > C.y) { C.y = col.y; }
+  if (col.z > C.z) { C.z = col.z; }
+  if (col.w > C.w) { C.w = col.w; }
```

Comment:
looking at this now, I think you may be able to do 
```
return max(col, C);
```
I believe glsl does a component-wise maximum that way and it has the possibility to be vectorized in the hardware.

Reply:
in that case, does it make sense to keep `accumulateMax` around at all or just write `C = max(col, C);` straight in the loop?

Reply:
yeah, you could put it directly in the loop without the wrapper function